### PR TITLE
 Move quote beta-reduction to QuoteContext

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/tastyreflect/KernelImpl.scala [hunk 1]
```diff
@@ -1894,6 +1895,33 @@ class KernelImpl(val rootContext: core.Contexts.Context) extends Kernel {
     case _ => None
   }
 
+  def betaReduce(fn: Term, args: List[Term]) given (ctx: Context): Term = {
+    val (argVals, argRefs) = args.map(arg => arg.tpe match {
+      case tpe: SingletonType if isIdempotentExpr(arg) => (Nil, arg)
+      case _ =>
+        val argVal = SyntheticValDef(NameKinds.UniqueName.fresh("x".toTermName), arg).withSpan(arg.span)
+        (argVal :: Nil, ref(argVal.symbol))
+    }).unzip
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/KernelImpl.scala]: It seems better to use `foldLeft`.

File: library/src/scala/tasty/reflect/Kernel.scala [hunk 2]
```diff
@@ -1538,4 +1538,6 @@ trait Kernel {
    */
   def searchImplicit(tpe: Type) given (ctx: Context): ImplicitSearchResult
 
+  def betaReduce(f: Term, args: List[Term]) given (ctx: Context): Term
```
Comment (liufengyun) [inline library/src/scala/tasty/reflect/Kernel.scala]: Maybe add some documentation here.

File: compiler/src/dotty/tools/dotc/tastyreflect/KernelImpl.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/KernelImpl.scala]: Changed

File: library/src/scala/tasty/reflect/Kernel.scala (hunk 2 above)
Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/Kernel.scala]: Added the doc in `Kernel.scala`

**Pull request merged**